src/Streamly/LZ4.hs

```diff
@@ -0,0 +1,106 @@
+-- |
+-- Module      : Streamly.LZ4
+-- Copyright   : (c) 2020 Composewell Technologies
+--
+-- License     : Apache-2.0
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- This module provides streamly combinators that perform LZ4
+-- compression. Please take a look at <https://github.com/lz4/lz4> to learn
+-- more. All combinators work with stream of 'Array' provided by streamly.
+--
+-- The important combinators are 'compress' and 'decompress'. Their behaviour is
+-- straightforward.
+--
+module Streamly.LZ4
+    ( compress
+    , resize
+    , decompressResized
+    , decompress
+    ) where
+
+--------------------------------------------------------------------------------
+-- Developer notes
+--------------------------------------------------------------------------------
+
+-- If you look at LZ4 as a black box, the idea behind this library is very
+-- simple.
+--
+-- During compression, all array elements are compressed and prefixed with 8
+-- bytes of meta data. 4 bytes for the compressed size 4 bytes for the
+-- decompressed.
+--
+-- If the compressed stream is ever saved to file then this meta data behaves
+-- like array boundaries. 'resize' uses this meta data to create a stream that
+-- respects these boundaries.
+--
+-- During decompression, the meta data is looked up and the compressed data used
+-- for decompression.
+
+--------------------------------------------------------------------------------
+-- Imports
+--------------------------------------------------------------------------------
+
+import Control.Monad.IO.Class (MonadIO)
+import Data.Word (Word8)
+import Streamly.Internal.Data.Array.Storable.Foreign (Array)
+import Streamly.Internal.Data.Stream.StreamD (fromStreamD, toStreamD)
+import Streamly.Prelude (SerialT)
+
+import Streamly.Internal.LZ4
+
+--------------------------------------------------------------------------------
+-- Compression
+--------------------------------------------------------------------------------
+
+-- | Compress a stream of foreign arrays (< 2GB / element) using LZ4 stream
+-- compression.
```

Comment:
If it does not work reliably for more than 2GB then we should check the length and error out if it is more than that.

Reply:
`lz4-conduit` has this in its documentation. I don't know why the foreign arrays need to be `< 2GB / element`. I'll add an assert for the time being but I think this can be removed.

Reply:
Its because the return type of `LZ4_compress_fast_continue` is 32-bit signed int which cannot store more than 2GB length.